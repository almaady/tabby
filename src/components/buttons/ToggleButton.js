import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Icon from "../icons/Icon";

class ToggleButton extends React.PureComponent {
  /**
   * @property {string} icon -The icon of the button
   * @property {func} onClick - The function triggered by the button
   * @property {string} backgroundColor -  The color of the background
   * @property {string} onLabel - Text to explain the "on" state
   * @property {bool} offLabel - Text to explain the "off" state
   * @property {bool} toggleState - The state of the button
   * @see See './../../../styles/variables/_colors. scss' for the list of color names
   */
  static propTypes = {
    icon: PropTypes.string,
    backgroundColor: PropTypes.string,
    onLabel: PropTypes.string,
    offLabel: PropTypes.string,
    toggleState: PropTypes.bool,
    onClick: PropTypes.func.isRequired
  };

  static defaultProps = {
    backgroundColor: "blue",
    onLabel: "",
    offLabel: "",
    icon: "",
    toggleState: true
  };

  state = {
    toggleState: this.props.toggleState
  };

  toggle = () => {
    const { toggleState } = this.state;
    const newState = !toggleState;
    this.setState({
      toggleState: newState
    });
    return newState;
  };

  onClickToggle = () => {
    const { onClick } = this.props;
    const actualState = this.toggle();
    onClick(actualState);
  };

  render() {
    const { icon, backgroundColor, onLabel, offLabel } = this.props;
    const { toggleState } = this.state;

    return (
      <div className="toggleButton--container">
        <p>{offLabel}</p>
        <div
          className={classNames("toggle--outside", `bg--${backgroundColor}`, {
            "toggle--off": !toggleState
          })}
        >
          <button
            type="button"
            className="toggle--inside"
            onClick={this.onClickToggle}
          >
            <Icon icon={icon} color={backgroundColor} size="R" />
          </button>
        </div>
        <p>{onLabel}</p>
      </div>
    );
  }
}

export default ToggleButton;
