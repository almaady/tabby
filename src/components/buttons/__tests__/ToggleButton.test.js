import React from "react"
import {shallow, mount} from "enzyme"
import sinon from "sinon"

import ToggleButton from "../ToggleButton"


describe("ToggleButton", () => {
  const action = sinon.stub();
  let props;
  beforeEach(() => {
    props = {
      backgroundColor:"blue",
      onLabel:"On text",
      offLabel:"Off text",
      icon:"user-outline",
      toggleState: true,
      onClick: action
    };
  });


  it("should render correctly with default state", () => {
    const component = shallow(<ToggleButton {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should change state when you click", () => {
    const component =  mount(<ToggleButton {...props} />)
    component.find(".toggle--inside").simulate("click");
    expect(component.state().toggleState).toBeFalse();

  });
  it("should add className toggle-off when the state is false", () => {
     const component = shallow(<ToggleButton {...props} />);
    component.find(".toggle--inside").simulate("click");
    expect(component.find(".toggle--outside").hasClass("toggle--off")).toBeTruthy();
  });
  it("should call action toggle-off when the state is false", () => {
    const component = shallow(<ToggleButton {...props} />);
    component.find("button").simulate("click");
    expect(action.called).toBeTruthy();
  });
});
