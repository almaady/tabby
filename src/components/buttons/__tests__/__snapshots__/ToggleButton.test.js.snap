// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ToggleButton
    backgroundColor="blue"
    icon="user-outline"
    offLabel="Off text"
    onClick={[Function]}
    onLabel="On text"
    toggleState={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
          Off text
        </p>,
        <div
          className="toggle--outside bg--blue"
        >
          <button
            className="toggle--inside"
            onClick={[Function]}
            type="button"
          >
            <Icon
              color="blue"
              icon="user-outline"
              size="R"
            />
          </button>
        </div>,
        <p>
          On text
        </p>,
      ],
      "className": "toggleButton--container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Off text",
        },
        "ref": null,
        "rendered": "Off text",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="toggle--inside"
            onClick={[Function]}
            type="button"
          >
            <Icon
              color="blue"
              icon="user-outline"
              size="R"
            />
          </button>,
          "className": "toggle--outside bg--blue",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Icon
              color="blue"
              icon="user-outline"
              size="R"
            />,
            "className": "toggle--inside",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "color": "blue",
              "icon": "user-outline",
              "size": "R",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "button",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "On text",
        },
        "ref": null,
        "rendered": "On text",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            Off text
          </p>,
          <div
            className="toggle--outside bg--blue"
          >
            <button
              className="toggle--inside"
              onClick={[Function]}
              type="button"
            >
              <Icon
                color="blue"
                icon="user-outline"
                size="R"
              />
            </button>
          </div>,
          <p>
            On text
          </p>,
        ],
        "className": "toggleButton--container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Off text",
          },
          "ref": null,
          "rendered": "Off text",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="toggle--inside"
              onClick={[Function]}
              type="button"
            >
              <Icon
                color="blue"
                icon="user-outline"
                size="R"
              />
            </button>,
            "className": "toggle--outside bg--blue",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Icon
                color="blue"
                icon="user-outline"
                size="R"
              />,
              "className": "toggle--inside",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "color": "blue",
                "icon": "user-outline",
                "size": "R",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "button",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "On text",
          },
          "ref": null,
          "rendered": "On text",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
